File: java/test/integration/cluster/FailoverTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.typedb.driver.test.integration.cluster;

import com.typedb.driver.TypeDB;
import com.typedb.driver.api.Credentials;
import com.typedb.driver.api.Driver;
import com.typedb.driver.api.DriverOptions;
import com.typedb.driver.api.DriverTlsConfig;
import com.typedb.driver.api.Transaction;
import com.typedb.driver.api.answer.QueryAnswer;
import com.typedb.driver.api.server.Server;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.util.Optional;
import java.util.Set;

import static java.util.Objects.requireNonNull;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class FailoverTest {

    private static final Set<String> ADDRESSES = Set.of(
            "127.0.0.1:11729", "127.0.0.1:21729", "127.0.0.1:31729"
    );
    private static final String USERNAME = "admin";
    private static final String PASSWORD = "password";
    private static final String DATABASE_NAME = "test-failover";
    private static final int FAILOVER_ITERATIONS = 10;
    private static final int PRIMARY_POLL_RETRIES = 20;
    private static final int PRIMARY_POLL_INTERVAL_SECS = 2;
    private static final int PRIMARY_FAILOVER_RETRIES = 5;
    private static final String CLUSTER_SERVER_SCRIPT = requireNonNull(
            System.getenv("CLUSTER_SERVER_SCRIPT"), "CLUSTER_SERVER_SCRIPT environment variable must be set"
    );

    private static void clusterServer(String command, String nodeId) throws IOException, InterruptedException {
        ProcessBuilder pb = new ProcessBuilder(CLUSTER_SERVER_SCRIPT, command, nodeId).inheritIO();
        if (System.getenv("CLUSTER_DIR") == null && System.getenv("BUILD_WORKSPACE_DIRECTORY") != null) {
            pb.environment().put("CLUSTER_DIR", System.getenv("BUILD_WORKSPACE_DIRECTORY"));
        }
        int exitCode = pb.start().waitFor();
        if (exitCode != 0) {
            fail(CLUSTER_SERVER_SCRIPT + " " + command + " " + nodeId + " failed with exit code " + exitCode);
        }
    }

    private static void ensureAllNodesUp() throws IOException, InterruptedException {
        for (int i = 1; i <= ADDRESSES.size(); i++) {
            clusterServer("start", String.valueOf(i));
            clusterServer("await", String.valueOf(i));
        }
    }

    private static String nodeIdFromAddress(String address) {
        String port = address.substring(address.lastIndexOf(':') + 1);
        return port.substring(0, 1);
    }

    private static Driver createDriver() throws InterruptedException {
        String rootCA = System.getenv("ROOT_CA");
        assertNotNull("ROOT_CA environment variable must be set", rootCA);
        for (int attempt = 0; attempt < PRIMARY_POLL_RETRIES; attempt++) {
            try {
                return TypeDB.driver(
                        ADDRESSES,
                        new Credentials(USERNAME, PASSWORD),
                        new DriverOptions(DriverTlsConfig.enabledWithRootCA(rootCA))
                                .primaryFailoverRetries(PRIMARY_FAILOVER_RETRIES)
                );
            } catch (Exception e) {
                if (attempt < PRIMARY_POLL_RETRIES - 1) {
                    System.out.println("  Driver creation failed (attempt " + (attempt + 1) + "/"
                            + PRIMARY_POLL_RETRIES + "): " + e.getMessage()
                            + ". Retrying in " + PRIMARY_POLL_INTERVAL_SECS + "s...");
                    Thread.sleep(PRIMARY_POLL_INTERVAL_SECS * 1000L);
                } else {
                    throw e;
                }
            }
        }
        throw new AssertionError("unreachable");
    }

    private static Server getPrimaryServer(Driver driver) throws InterruptedException {
        for (int attempt = 0; attempt < PRIMARY_POLL_RETRIES; attempt++) {
            Optional<? extends Server> primary = driver.primaryServer();
            if (primary.isPresent()) {
                return primary.get();
            }
            if (attempt < PRIMARY_POLL_RETRIES - 1) {
                System.out.println("  No primary server found (attempt " + (attempt + 1) + "/" + PRIMARY_POLL_RETRIES
                        + "). Retrying in " + PRIMARY_POLL_INTERVAL_SECS + "s...");
                Thread.sleep(PRIMARY_POLL_INTERVAL_SECS * 1000L);
            }
        }
        fail("Retry limit exceeded while seeking a primary server.");
        return null;
    }

    private static void setupDatabase(Driver driver) throws InterruptedException {
        for (int attempt = 0; attempt < PRIMARY_POLL_RETRIES; attempt++) {
            try {
                if (driver.databases().contains(DATABASE_NAME)) {
                    driver.databases().get(DATABASE_NAME).delete();
                }
                driver.databases().create(DATABASE_NAME);
                try (Transaction tx = driver.transaction(DATABASE_NAME, Transaction.Type.SCHEMA)) {
                    tx.query("define entity person;").resolve();
                    tx.commit();
                }
                return;
            } catch (Exception e) {
                if (attempt < PRIMARY_POLL_RETRIES - 1) {
                    System.out.println("  Database setup failed (attempt " + (attempt + 1) + "/"
                            + PRIMARY_POLL_RETRIES + "): " + e.getMessage()
                            + ". Retrying in " + PRIMARY_POLL_INTERVAL_SECS + "s...");
                    Thread.sleep(PRIMARY_POLL_INTERVAL_SECS * 1000L);
                } else {
                    throw e;
                }
            }
        }
    }

    @Before
    public void setUp() throws IOException, InterruptedException {
        ensureAllNodesUp();
        try {
            Driver driver = createDriver();
            if (driver.databases().contains(DATABASE_NAME)) {
                driver.databases().get(DATABASE_NAME).delete();
            }
            driver.close();
        } catch (Exception e) {
            // Best effort cleanup
        }
    }

    @After
    public void tearDown() {
        try {
            Driver driver = createDriver();
            if (driver.databases().contains(DATABASE_NAME)) {
                driver.databases().get(DATABASE_NAME).delete();
            }
            driver.close();
        } catch (Exception e) {
            // Best effort cleanup
        }
    }

    @Test
    public void primaryFailover() throws Exception {
        System.out.println("=== Cluster Failover Test ===");

        System.out.println("Connecting driver...");
        Driver driver = createDriver();

        System.out.println("Setting up database and schema...");
        setupDatabase(driver);
        verifyReadQuery(driver);
        System.out.println("Initial setup verified.");

        for (int iteration = 1; iteration <= FAILOVER_ITERATIONS; iteration++) {
            System.out.println("\n--- Failover iteration " + iteration + "/" + FAILOVER_ITERATIONS + " ---");

            Server primary = getPrimaryServer(driver);
            assertNotNull(primary);
            String primaryAddress = primary.getAddress();
            String nodeId = nodeIdFromAddress(primaryAddress);
            System.out.println("  Primary server: " + primaryAddress + " (node " + nodeId + ")");

            System.out.println("  Read query before kill...");
            verifyReadQuery(driver);

            System.out.println("  Killing node " + nodeId + "...");
            clusterServer("kill", nodeId);

            System.out.println("  Read query immediately after kill (driver auto-failover)...");
            verifyReadQuery(driver);
            System.out.println("  Auto-failover read succeeded.");

            System.out.println("  Confirming new primary...");
            Server newPrimary = getPrimaryServer(driver);
            System.out.println("  New primary: " + newPrimary.getAddress()
                    + " (node " + nodeIdFromAddress(newPrimary.getAddress()) + ")");

            System.out.println("  Read query on confirmed primary...");
            verifyReadQuery(driver);
            System.out.println("  Confirmed primary read succeeded.");

            System.out.println("  Restarting node " + nodeId + "...");
            clusterServer("start", nodeId);
            clusterServer("await", nodeId);
            System.out.println("  Node " + nodeId + " restarted.");
        }

        System.out.println("\n=== All " + FAILOVER_ITERATIONS + " failover iterations passed! ===");
        driver.close();
    }

    private void verifyReadQuery(Driver driver) {
        try (Transaction tx = driver.transaction(DATABASE_NAME, Transaction.Type.READ)) {
            QueryAnswer answer = tx.query("match entity $t;").resolve();
            assertTrue("Expected at least one entity type in read query results",
                    answer.asConceptRows().stream().findAny().isPresent());
        }
    }
}

File: csharp/Test/Integration/FailoverTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using NUnit.Framework;
using TypeDB.Driver;
using TypeDB.Driver.Api;
using TypeDB.Driver.Common;

namespace TypeDB.Driver.Test.Integration
{
    [TestFixture]
    public class FailoverTest
    {
        private static readonly ISet<string> Addresses = new HashSet<string>
        {
            "127.0.0.1:11729", "127.0.0.1:21729", "127.0.0.1:31729"
        };
        private const string Username = "admin";
        private const string Password = "password";
        private const string DatabaseName = "test-failover";
        private const int FailoverIterations = 10;
        private const int PrimaryPollRetries = 20;
        private const int PrimaryPollIntervalSecs = 2;
        private const int PrimaryFailoverRetries = 5;
        private static readonly string ClusterServerScript =
            Environment.GetEnvironmentVariable("CLUSTER_SERVER_SCRIPT")
            ?? throw new InvalidOperationException("CLUSTER_SERVER_SCRIPT environment variable must be set");

        private static void ClusterServer(string command, string nodeId)
        {
            var process = new Process();
            process.StartInfo.FileName = ClusterServerScript;
            process.StartInfo.Arguments = $"{command} {nodeId}";
            process.StartInfo.UseShellExecute = false;
            process.StartInfo.RedirectStandardError = true;
            if (Environment.GetEnvironmentVariable("CLUSTER_DIR") == null
                && Environment.GetEnvironmentVariable("BUILD_WORKSPACE_DIRECTORY") != null)
            {
                process.StartInfo.EnvironmentVariables["CLUSTER_DIR"] =
                    Environment.GetEnvironmentVariable("BUILD_WORKSPACE_DIRECTORY");
            }
            process.Start();
            var stderr = process.StandardError.ReadToEnd();
            process.WaitForExit();
            Assert.AreEqual(0, process.ExitCode,
                $"{ClusterServerScript} {command} {nodeId} failed: {stderr}");
        }

        private static void EnsureAllNodesUp()
        {
            for (int i = 1; i <= Addresses.Count; i++)
            {
                ClusterServer("start", i.ToString());
                ClusterServer("await", i.ToString());
            }
        }

        private static string NodeIdFromAddress(string address)
        {
            var port = address.Substring(address.LastIndexOf(':') + 1);
            return port.Substring(0, 1);
        }

        private static IDriver CreateDriver()
        {
            var rootCA = Environment.GetEnvironmentVariable("ROOT_CA");
            Assert.IsNotNull(rootCA, "ROOT_CA environment variable must be set");
            for (int attempt = 0; attempt < PrimaryPollRetries; attempt++)
            {
                try
                {
                    var options = new DriverOptions(DriverTlsConfig.EnabledWithRootCA(rootCA));
                    options.PrimaryFailoverRetries = PrimaryFailoverRetries;
                    return TypeDB.Driver(Addresses, new Credentials(Username, Password), options);
                }
                catch (Exception e)
                {
                    if (attempt < PrimaryPollRetries - 1)
                    {
                        Console.WriteLine($"  Driver creation failed (attempt {attempt + 1}/{PrimaryPollRetries}): "
                            + $"{e.Message}. Retrying in {PrimaryPollIntervalSecs}s...");
                        Thread.Sleep(PrimaryPollIntervalSecs * 1000);
                    }
                    else
                    {
                        throw;
                    }
                }
            }
            throw new InvalidOperationException("unreachable");
        }

        private static IServer GetPrimaryServer(IDriver driver)
        {
            for (int attempt = 0; attempt < PrimaryPollRetries; attempt++)
            {
                var primary = driver.GetPrimaryServer();
                if (primary != null)
                {
                    return primary;
                }
                if (attempt < PrimaryPollRetries - 1)
                {
                    Console.WriteLine($"  No primary server found (attempt {attempt + 1}/{PrimaryPollRetries}). "
                        + $"Retrying in {PrimaryPollIntervalSecs}s...");
                    Thread.Sleep(PrimaryPollIntervalSecs * 1000);
                }
            }
            Assert.Fail("Retry limit exceeded while seeking a primary server.");
            return null;
        }

        private static void SetupDatabase(IDriver driver)
        {
            for (int attempt = 0; attempt < PrimaryPollRetries; attempt++)
            {
                try
                {
                    if (driver.Databases.Contains(DatabaseName))
                    {
                        driver.Databases.Get(DatabaseName).Delete();
                    }
                    driver.Databases.Create(DatabaseName);
                    using (var tx = driver.Transaction(DatabaseName, TransactionType.Schema))
                    {
                        tx.Query("define entity person;").Resolve();
                        tx.Commit();
                    }
                    return;
                }
                catch (Exception e)
                {
                    if (attempt < PrimaryPollRetries - 1)
                    {
                        Console.WriteLine($"  Database setup failed (attempt {attempt + 1}/{PrimaryPollRetries}): "
                            + $"{e.Message}. Retrying in {PrimaryPollIntervalSecs}s...");
                        Thread.Sleep(PrimaryPollIntervalSecs * 1000);
                    }
                    else
                    {
                        throw;
                    }
                }
            }
        }

        [SetUp]
        public void SetUp()
        {
            EnsureAllNodesUp();
            try
            {
                using var driver = CreateDriver();
                if (driver.Databases.Contains(DatabaseName))
                {
                    driver.Databases.Get(DatabaseName).Delete();
                }
            }
            catch (Exception)
            {
                // Best effort cleanup
            }
        }

        [TearDown]
        public void TearDown()
        {
            try
            {
                using var driver = CreateDriver();
                if (driver.Databases.Contains(DatabaseName))
                {
                    driver.Databases.Get(DatabaseName).Delete();
                }
            }
            catch (Exception)
            {
                // Best effort cleanup
            }
        }

        [Test]
        public void PrimaryFailover()
        {
            Console.WriteLine("=== Cluster Failover Test ===");

            Console.WriteLine("Connecting driver...");
            using var driver = CreateDriver();

            Console.WriteLine("Setting up database and schema...");
            SetupDatabase(driver);
            VerifyReadQuery(driver);
            Console.WriteLine("Initial setup verified.");

            for (int iteration = 1; iteration <= FailoverIterations; iteration++)
            {
                Console.WriteLine($"\n--- Failover iteration {iteration}/{FailoverIterations} ---");

                var primary = GetPrimaryServer(driver);
                Assert.IsNotNull(primary);
                var primaryAddress = primary.Address;
                var nodeId = NodeIdFromAddress(primaryAddress);
                Console.WriteLine($"  Primary server: {primaryAddress} (node {nodeId})");

                Console.WriteLine("  Read query before kill...");
                VerifyReadQuery(driver);

                Console.WriteLine($"  Killing node {nodeId}...");
                ClusterServer("kill", nodeId);

                Console.WriteLine("  Read query immediately after kill (driver auto-failover)...");
                VerifyReadQuery(driver);
                Console.WriteLine("  Auto-failover read succeeded.");

                Console.WriteLine("  Confirming new primary...");
                var newPrimary = GetPrimaryServer(driver);
                Console.WriteLine($"  New primary: {newPrimary.Address} "
                    + $"(node {NodeIdFromAddress(newPrimary.Address)})");

                Console.WriteLine("  Read query on confirmed primary...");
                VerifyReadQuery(driver);
                Console.WriteLine("  Confirmed primary read succeeded.");

                Console.WriteLine($"  Restarting node {nodeId}...");
                ClusterServer("start", nodeId);
                ClusterServer("await", nodeId);
                Console.WriteLine($"  Node {nodeId} restarted.");
            }

            Console.WriteLine($"\n=== All {FailoverIterations} failover iterations passed! ===");
        }

        private void VerifyReadQuery(IDriver driver)
        {
            using var tx = driver.Transaction(DatabaseName, TransactionType.Read);
            var answer = tx.Query("match entity $t;").Resolve()!;
            var rows = answer.AsConceptRows();
            Assert.IsTrue(rows.Any(), "Expected at least one entity type in read query results");
        }
    }
}
